File: sdks/browser4-sdk-python/examples/basic_usage.py
"""
Basic usage example for Browser4 Python SDK.

This example demonstrates fundamental Browser4 SDK operations:
- Using Browser4Driver to automatically start the server (recommended)
- Creating a client and session
- Loading pages
- Navigating with WebDriver
- Extracting data with CSS selectors
- AI-powered actions

Note: This example uses Browser4Driver which automatically downloads
      and starts the Browser4 server. No manual setup required!
"""

from browser4 import Browser4Driver, PulsarClient, AgenticSession

def main():
    print("=== Browser4 Python SDK - Basic Usage Example ===\n")

    # Use Browser4Driver to automatically start the server
    print("Starting Browser4 server...")
    driver = Browser4Driver()

    try:
        driver.start()
        print(f"Server started at: {driver.base_url}\n")

        # Create client and session
        print("Creating session...")
        client = PulsarClient(base_url=driver.base_url)
    except:
        print("Failed to start Browser4 server.")
        return

    try:
        session_id = client.create_session()
        print(f"Session created: {session_id}\n")

        session = AgenticSession(client)

        # Load a page
        print("Loading example.com...")
        page = session.open("https://example.com")
        print(f"Page loaded: {page.url}")
        print(f"Content type: {page.content_type}")
        print(f"Content length: {page.content_length} bytes\n")

        # Parse the page
        document = session.parse(page)

        # Extract data using CSS selectors
        if document:
            print("Extracting data with CSS selectors...")
            fields = session.extract(document, {
                "title": "h1",
                "description": "p"
            })

            print("Extracted data:")
            for key, value in fields.items():
                print(f"  {key}: {value[:100] if value else 'None'}...")
            print()

        # Use WebDriver for navigation
        driver = session.driver
        print("Getting current URL from WebDriver...")
        current_url = driver.current_url()
        print(f"Current URL: {current_url}")
        print(f"Page title: {driver.title()}\n")

        # AI-powered action (requires AI capabilities)
        try:
            print("Attempting AI-powered action...")
            result = session.act("scroll to the bottom of the page")
            print(f"Action success: {result.success}")
            print(f"Action message: {result.message}\n")
        except Exception as e:
            print(f"AI action skipped: {e}\n")

        # Clean up
        print("Closing session...")
        session.close()
        print("Session closed successfully!")

    except Exception as e:
        print(f"Error: {e}")
    finally:
        client.close()
        driver.stop()

    print("\n=== Example completed ===")
    print("\nNote: On first run, Browser4.jar is downloaded automatically.")
    print("      This is a one-time operation and may take a few minutes.")


if __name__ == "__main__":
    main()

File: sdks/browser4-nodejs/examples/basic-usage.ts
/**
 * Basic example demonstrating Browser4 NodeJS SDK usage.
 * 
 * This example shows how to:
 * - Start Browser4 server automatically
 * - Create a session
 * - Navigate and interact with pages
 * - Use AI-powered actions
 */

import { Browser4Driver, PulsarClient, AgenticSession } from '../src';

async function main() {
  console.log('Starting Browser4 server...');
  
  // Browser4Driver automatically downloads and starts the server
  const driver = new Browser4Driver();
  
  await driver.use(async (d) => {
    console.log(`Server running at: ${d.baseUrl}`);
    
    // Create client and session
    const client = new PulsarClient({ baseUrl: d.baseUrl });
    const sessionId = await client.createSession();
    console.log(`Session created: ${sessionId}`);
    
    const session = new AgenticSession(client);
    
    // Navigate to a page
    console.log('\n=== Opening page ===');
    const page = await session.open('https://example.com');
    console.log(`Opened: ${page.url}`);
    console.log(`Content type: ${page.contentType}`);
    console.log(`Content length: ${page.contentLength} bytes`);
    
    // Use WebDriver for element interaction
    console.log('\n=== Using WebDriver ===');
    const webDriver = session.driver;
    
    // Get current URL
    const currentUrl = await webDriver.currentUrl();
    console.log(`Current URL: ${currentUrl}`);
    
    // Get page title
    const title = await webDriver.title();
    console.log(`Page title: ${title}`);
    
    // Check if an element exists
    const h1Exists = await webDriver.exists('h1');
    console.log(`H1 element exists: ${h1Exists}`);
    
    // Get text from an element
    if (h1Exists) {
      const h1Text = await webDriver.getText('h1');
      console.log(`H1 text: ${h1Text}`);
    }
    
    // Use AI-powered actions
    console.log('\n=== Using AI-powered actions ===');
    
    // Observe the page
    const observations = await session.observe('What can I interact with?');
    console.log(`Found ${observations.observations.length} observations`);
    observations.observations.slice(0, 3).forEach((obs, i) => {
      console.log(`  ${i + 1}. ${obs.method}: ${obs.description}`);
    });
    
    // Summarize the page
    const summary = await session.summarize('Summarize this page in one sentence');
    console.log(`Summary: ${summary}`);
    
    // Execute a single action
    const actResult = await session.act('scroll down 200 pixels');
    console.log(`Action result: ${actResult.success ? 'Success' : 'Failed'}`);
    console.log(`Message: ${actResult.message}`);
    
    // Check state history
    const history = session.stateHistory;
    console.log(`\nExecuted ${history.states.length} actions`);
    history.states.forEach((state, i) => {
      console.log(`  ${i + 1}. ${state.action}: ${state.success ? 'Success' : 'Failed'}`);
    });
    
    // Clean up
    console.log('\n=== Cleaning up ===');
    await session.close();
    client.close();
    console.log('Session closed');
  });
  
  console.log('Server stopped');
}

// Run the example
main().catch(error => {
  console.error('Error:', error);
  process.exit(1);
});
